dynamic_validators/insecure_dynamic_validators.md

```diff
@@ -0,0 +1,128 @@
+# Dynamically add/remove validators at runtime (Insecure version)
+
+## Problem Description
+In the current state of tendermint bdb, validator nodes are specified in the Tendermint config file before starting the node. It is necessary that the users should be able to dynamically add new validators.
+
+**NOTE**: This document presents an insecure implementation intended for bigchaindb tendermint MVP.
+
+### Technical details
+Tendermint allows the abci client application to return a list of validators during `end_block` call. If the client doesn't wish to modify the validators an empty list should be returned. The return value allows the client to dynamically add/remove validators. Note that the return value could be thought of as a (validators) diff of desired validators list and current validators list. Below is an example of validators diff,
+```json
+[{"pub_key":{"type":"ed25519","data":"4E2685D9016126864733225BE00F005515200727FBAB1312FC78C8B76831255A"},
+  "power":10},
+ {"pub_key":{"type":"ed25519","data":"608D839D7100466D6BA6BE79C320F8B81DE93CFAA58CF9768CF921C6371F2553"},
+ "power":0}
+ ]
+```
+Note: `"power": 0` implies that the validator should be removed from the validator list. In case this diff is empty the validators remain same. Furthermore, the validators returned during `end_block` are used to update the validators of the given Tendermint node and **not** for the whole network.
+
+
+## Proposed Change
+- Add a new subcommand `bighciandb upsert-validator PUBLIC_KEY POWER` which will allow the node admin to add/update/delete a validator.
+
+- `GET /api/v1/validators` should be introduced to list the current validators.
```

Comment:
I was re-reading this spec just now and realized I don't know if this `GET /api/v1/validators` request is supposed to respond with this node's _local_ list of validators, or some kind of network-wide agreed-upon list of validators (i.e. a list that >2/3 of the nodes agree on). I suspect it's the local list. Is that right?

Reply:
Yes, its the local list. I think by network-wide agreed-upon validators you mean the set of validators accepted by majority of other validators. In my view this information should be gathered by the client. This is similar to the scenario wherein a client needs to verify that a transaction was commited so it consults f+1 nodes.